### PR TITLE
Target .NET Framework

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/grpc/grpc-dotnet/issues/1176

Experiment targeting .NET Framework with Grpc.Net.Client

~Blocked on [WinHttpHandler: Read HTTP/2 trailing headers](https://github.com/dotnet/runtime/pull/46602).~ Key/type is unlikely to change. If it does then we can react relatively easily.

### CODE REVIEW

**Review by JunTaoLuo — COMMENTED** [review]
So far looks good, some minor comments. I have one more file to review in the morning.

**Review by JunTaoLuo — APPROVED** [review]

File: src/Grpc.Net.Client/Internal/GrpcProtocolConstants.cs
Comment (JamesNK) [reply src/Grpc.Net.Client/Internal/GrpcProtocolConstants.cs:66]: Yeah

File: test/Grpc.Net.Client.Tests/Grpc.Net.Client.Tests.csproj
Comment (JamesNK) [reply test/Grpc.Net.Client.Tests/Grpc.Net.Client.Tests.csproj:4]: `net4.7.2` works. Is it important?

File: test/Grpc.Net.Client.Web.Tests/GrpcWebResponseStreamTests.cs
Comment (JamesNK) [reply test/Grpc.Net.Client.Web.Tests/GrpcWebResponseStreamTests.cs:61]: It pulled in a source file that then created other conflicts. Using plain ol string instead solved a bunch of issues.

File: src/Grpc.Net.Client/Internal/CompatibilityExtensions.cs
Comment (JamesNK) [reply src/Grpc.Net.Client/Internal/CompatibilityExtensions.cs:18]: Good idea, but this check is fine because we only want the else in not .NET Standard 2.0 which won't change.

File: test/Grpc.Net.Client.Tests/Grpc.Net.Client.Tests.csproj
Comment (JunTaoLuo) [reply test/Grpc.Net.Client.Tests/Grpc.Net.Client.Tests.csproj:4]: I'm kind of surprised that this works. It's certainly not "canon" given how dotnet/runtime and dotnet/sdk uses 472 instead: https://github.com/dotnet/sdk/search?l=XML&p=3&q=net472. I did a quick search and didn't see anywhere in our stack that we use net4.7.2 but maybe the logic is flexible enough to handle it? Still, I don't think it's a good idea to deviate from norms.

Comment (JamesNK) [reply test/Grpc.Net.Client.Tests/Grpc.Net.Client.Tests.csproj]: Ok. Updated.

**Comment by JamesNK** [comment]
@jtattermusch Ready to review

If no HttpHandler/HttpClient is configured then it fails with a user friendly error. Once there is a compadible version of WinHttpHandler available (will come with .NET 6) then:

1. A .NET Framework target can be added that automatically wires up WinHttpHandler
2. It throws an error if your Windows version isn't new enough

File: testassets/InteropTestsClient/RunGrpcTests.ps1
Comment (jtattermusch) [inline testassets/InteropTestsClient/RunGrpcTests.ps1:46]: isn't it dangerous to drop tests with --use_tls=true?

File: test/Grpc.Net.Client.Tests/GrpcChannelTests.cs
Comment (jtattermusch) [inline test/Grpc.Net.Client.Tests/GrpcChannelTests.cs:165]: nit: the link points here: https://docs.microsoft.com/en-us/aspnet/core/grpc/basics?view=aspnetcore-5.0 is that indended?

**Review by jtattermusch — COMMENTED** [review]
This is looking pretty good. Had a few last questions before I approve.

**Review by jtattermusch — APPROVED** [review]
Initially I was concerned if adding the net472 target is going to make the solution hard to build and test linux (under mono), so I tested that on my linux workstation and the results are  looking good:
- with dotnet SDK 5 installed, it is possible to build the entire solution even without mono being installed.
- after installing mono (I have mono 6.8), it is possible to run the net472 test projects.

Looks like this is good to go (please look at my remaining comments) and thanks for making this possible!

One more ask: can you update the PR description to contain a summary of what's being done and what are the key results? (what features work on what framework / os is going to become a bit more interesting after this PR, so it's worth having a good PR description).

File: test/Grpc.Net.Client.Tests/GrpcChannelTests.cs
Comment (JamesNK) [reply test/Grpc.Net.Client.Tests/GrpcChannelTests.cs:165]: Its a temporary link until docs are added. That will happen when this feature is released.

Reminder issue - https://github.com/grpc/grpc-dotnet/issues/1200

File: testassets/InteropTestsClient/RunGrpcTests.ps1
Comment (JamesNK) [reply testassets/InteropTestsClient/RunGrpcTests.ps1:46]: Using a cert with test.google.com never successfully worked so I removed the if/else.

This file is just a helper for running all the interop tests locally and isn't used by CI servers.

**Comment by jtattermusch** [comment]
Question about this not being blocked on https://github.com/dotnet/runtime/pull/46602 anymore (see updated description) - does that mean that we no longer depend on making a change in WinHttpHandler to make unary and server streaming calls work? (and they will work with the lastest stable version of https://www.nuget.org/packages/System.Net.Http.WinHttpHandler/5.0.0).

- since the story of what works where and which versions of WinHttpHandler / Windows / other components are required, it would be useful to have a summary.

**Comment by JamesNK** [comment]
> Question about this PR not being blocked on [dotnet/runtime#46602](https://github.com/dotnet/runtime/pull/46602) anymore (see updated description) - does that mean that we no longer depend on making a change in WinHttpHandler to make unary and server streaming calls work? (and they will work with the lastest stable version of [nuget.org/packages/System.Net.Http.WinHttpHandler/5.0.0](https://www.nuget.org/packages/System.Net.Http.WinHttpHandler/5.0.0)).

WinHttpHandler changes are still needed. I double checked with the team for WinHttpHandler that the key/type for headers won't change, which is why I originally marked this as blocked. Because they won't change this PR can be merged and when WinHttpHandler vNext is released it can be plugged in and start working.

**Comment by jtattermusch** [comment]
What is the ETA for merging https://github.com/dotnet/runtime/pull/46602 and releasing a pre-release nuget then?
Without that we won't have trailers support, which makes grpc-dotnet over WinHttpHandler useless. We'd like to be able to test unary and server streaming calls over WinHttpHandler ASAP.

**Comment by jtattermusch** [comment]
> Fixes #1176
> 
> Experiment targeting .NET Framework with Grpc.Net.Client
> 
> ~Blocked on [WinHttpHandler: Read HTTP/2 trailing headers](https://github.com/dotnet/runtime/pull/46602).~ Key/type is unlikely to change. If it does then we can react relatively easily.
> 
> This change adds a `netstandard2.0` target to Grpc.Net.Client. That allows Grpc.Net.Client to be used on .NET Framework (and some other places).
> 
> Note that .NET Framework doesn't have built-in support for HTTP/2 so a custom HTTP handler needs to be configured. To use Grpc.Net.Client successfully on .NET Framework you must either use gRPC-Web, or set `GrpcChannelOptions.HttpHandler` to `WinHttpHandler`. That handler uses WinHttp, a HTTP client built into Windows, which supports HTTP/2. There are required versions and restrictions when doing this:
> 
> * Only unary and server streaming calls work in the short term. Longer term the goal is to support all gRPC calls.
> * An insiders build of Windows 10 is required to have a version of WinHttp that supports gRPC.

insiders build of windows is only required for client streaming and bidi streaming, right? 

> * A preview versions of WinHttpHandler for .NET 6 is required to use the new features in WinHttp **NOT YET DONE**

Just to clarify:  By saying "WinHttpHandler for .NET 6" you mean one of the future 6.0.0-previewXYZ versions  of that nuget package (the one that contains dotnet/runtime#46602)? For our purposes (using grpc-dotnet on .NET framework) the ".NET 6" part isn't really important is this is not where we're going to use it.

**Pull request merged**
